experiment/contractNipopow.sol

```diff
@@ -398,31 +254,99 @@ contract Crosschain {
         return true;
     }
 
+    // This will may be expensive. Check if a memory mapping is required
+    // Check if all blocks of existing[lca+1:] are different from contesting[1:]
+    function allDifferent(
+        bytes32[] memory existing,
+        bytes32[] memory contesting,
+        uint256 lca
+    ) internal pure returns (bool) {
+        for (uint256 i = lca + 1; i < existing.length; i++) {
+            for (uint256 j = 1; j < contesting.length; j++) {
+                if (existing[i] == contesting[j]) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
     function submitContestingProof(
-        bytes32[4][] memory headers,
-        bytes32[] memory siblings,
-        bytes32[4] memory blockOfInterest
+        bytes32[4][] memory existingHeaders,
+        uint256 lca,
+        bytes32[4][] memory contestingHeaders,
+        bytes32[] memory contestingSiblings,
+        uint256 blockOfInterestIndex
     ) public returns (bool) {
-        bytes32 hashedBlock = hashHeader(blockOfInterest);
-
-        if (events[hashedBlock].expire <= block.number) {
-            return false;
+        require(
+            existingHeaders.length > blockOfInterestIndex &&
+                blockOfInterestIndex >= 0,
+            "Block of interest index is out of range"
+        );
+
+        bytes32 blockOfInterestHash = hashHeader(
+            existingHeaders[blockOfInterestIndex]
+        );
+
+        require(
+            events[blockOfInterestHash].expire > block.number,
+            "Contesting period has expired"
+        );
+
+        require(existingHeaders.length > lca, "Lca out of range");
+
+        require(
+            lca > blockOfInterestIndex,
+            "Block of interest exists in sub-chain"
+        );
+
+        require(
+            events[blockOfInterestHash].proofHash == hashProof(existingHeaders),
+            "Wrong existing proof"
+        );
+
+        // get contesting hashed headers
+        bytes32[] memory contestingHeadersHashed = new bytes32[](
+            contestingHeaders.length
+        );
+        for (uint256 i = 0; i < contestingHeaders.length; i++) {
+            contestingHeadersHashed[i] = hashHeader(contestingHeaders[i]);
         }
-
-        if (
-            !verify(
-                events[hashedBlock].proof,
-                headers,
-                siblings,
-                blockOfInterest
-            )
-        ) {
-            events[hashedBlock].expire = 0;
-            msg.sender.transfer(z);
-            return true;
+        validateInterlink(
+            contestingHeaders,
+            contestingHeadersHashed,
+            contestingSiblings
+        );
+
+        // get existing hashed headers
```

Comment:
nit: unnecessary comment

Reply:
repetitive code of getting the header hashes, can you extract it?

Reply:
It can be done, but it costs more gas to create the array inside a function and return it. And there is no way to have a global memory array.

Reply:
how much more?